File: 2015/Day23/main.dart
import 'dart:io';

void main() async {
  var instructions = await readInstructions('input.txt');
  print('Part 1: ${executeProgram(instructions)}');
  print('Part 2: ${executeProgram(instructions, part2: true)}');
}

int executeProgram(List<Instruction> instructions, {bool part2 = false}) {
  final registers = {'a': part2 ? 1 : 0, 'b': 0};
  var pc = 0;

  while (pc >= 0 && pc < instructions.length) {
    final instr = instructions[pc];

    if (instr.op == 'hlf') {
      registers[instr.reg] = (registers[instr.reg]! / 2).floor();
      pc++;
    } else if (instr.op == 'tpl') {
      registers[instr.reg] = registers[instr.reg]! * 3;
      pc++;
    } else if (instr.op == 'inc') {
      registers[instr.reg] = registers[instr.reg]! + 1;
      pc++;
    } else if (instr.op == 'jmp') {
      pc += instr.offset;
    } else if (instr.op == 'jie') {
      if (registers[instr.reg]! % 2 == 0) {
        pc += instr.offset;
      } else {
        pc++;
      }
    } else if (instr.op == 'jio') {
      if (registers[instr.reg] == 1) {
        pc += instr.offset;
      } else {
        pc++;
      }
    }
  }

  return registers['b']!;
}

Future<List<Instruction>> readInstructions(String filename) async {
  final List<Instruction> instructions = [];
  final lines = await File(filename).readAsLines();

  for (var line in lines) {
    final parts = line.split(' ');
    final op = parts[0];
    var reg = '';
    var offset = 0;

    if (op == 'hlf' || op == 'tpl' || op == 'inc') {
      reg = parts[1];
    } else if (op == 'jmp') {
      offset = int.parse(parts[1]);
    } else if (op == 'jie' || op == 'jio') {
      reg = parts[1].substring(0, parts[1].length - 1);
      offset = int.parse(parts[2]);
    }

    instructions.add(Instruction(op, reg, offset));
  }

  return instructions;
}

class Instruction {
  String op;
  String reg;
  int offset;

  Instruction(this.op, this.reg, this.offset);
}

File: 2015/Day23/main.lua
local function read_instructions(filename)
    local instructions = {}
    for line in io.lines(filename) do
        local parts = {}
        for part in string.gmatch(line, "%S+") do
            table.insert(parts, part)
        end

        local op = parts[1]
        local reg = ''
        local offset = 0

        if op == 'hlf' or op == 'tpl' or op == 'inc' then
            reg = parts[2]
        elseif op == 'jmp' then
            offset = tonumber(parts[2]) or 0
        elseif op == 'jie' or op == 'jio' then
            reg = string.sub(parts[2], 1, -2)
            offset = tonumber(parts[3]) or 0
        end

        table.insert(instructions, { op = op, reg = reg, offset = offset })
    end
    return instructions
end

local function execute_program(instructions, part2)
    part2 = part2 or false
    local registers = { a = part2 and 1 or 0, b = 0 }
    local pc = 0

    while pc >= 0 and pc < #instructions do
        local instr = instructions[pc + 1]

        if instr.op == 'hlf' then
            registers[instr.reg] = math.floor(registers[instr.reg] / 2)
            pc = pc + 1
        elseif instr.op == 'tpl' then
            registers[instr.reg] = registers[instr.reg] * 3
            pc = pc + 1
        elseif instr.op == 'inc' then
            registers[instr.reg] = registers[instr.reg] + 1
            pc = pc + 1
        elseif instr.op == 'jmp' then
            pc = pc + instr.offset
        elseif instr.op == 'jie' then
            if registers[instr.reg] % 2 == 0 then
                pc = pc + instr.offset
            else
                pc = pc + 1
            end
        elseif instr.op == 'jio' then
            if registers[instr.reg] == 1 then
                pc = pc + instr.offset
            else
                pc = pc + 1
            end
        end
    end

    return registers.b
end

local instructions = read_instructions('input.txt')
print('Part 1:', execute_program(instructions))
print('Part 2:', execute_program(instructions, true))
